### PR TITLE
Ignore errors based on policy settings

### PR DESCRIPTION
**Pull request by redgoat650**
Ignore errors on directory and file reads based on policy. Still need to add more testing. Snapshot fail test updated to mirror current behavior without changes to the test cases (set policy to ignore file errors).

### CODE REVIEW

**Comment by redgoat650** [comment]
PR follows #34

**Comment by redgoat650** [comment]
Hold off on reviewing this one, something's not right in a few corner cases (caught in the SnapshotFail test).

**Comment by redgoat650** [comment]
Okay I think this is ready again for review @pavannd1 @julio-lopez. Please pay special attention to the test case expected outcomes when you review, make sure you agree with the expected behavior.

I have implemented it such that the policy ignores read errors ONLY on child entries of the source. So a snapshot will still fail if the source root directory itself can't be read, but you can ignore the error if a file or a subdirectory in the snapshot source root can't be read. I did this to address some otherwise strange behavior where you would successfully snapshot (because you ignored the error), but couldn't restore that snapshot because nothing really happened during the operation.

**Comment by redgoat650** [comment]
I'd recommend viewing changes as https://github.com/kastenhq/kopia/pull/35/files?w=1, since most of the default view changes are from whitespace

File: snapshot/snapshotfs/upload.go [hunk 1]
```diff
@@ -606,6 +614,12 @@ func uniqueDirectories(dirs []fs.Directory) []fs.Directory {
 	return result
 }
 
+// DirReadError distinguishes an error thrown when attempting to
+// read a directory
+type DirReadError struct {
```
Comment (julio-lopez) [inline snapshot/snapshotfs/upload.go]: Could this be changed to a type redefinition and still have the type assertion above work unmodified?
```suggestion
type DirReadError error
```

File: snapshot/snapshotfs/upload.go [hunk 2]
```diff
@@ -39,6 +39,9 @@ type Uploader struct {
 	// ignore file read errors
 	IgnoreFileErrors bool
 
+	// ignore directory read errors
+	IgnoreDirErrors bool
```
Comment (julio-lopez) [inline snapshot/snapshotfs/upload.go]: Does it make sense to continue having these fields in the `Uploader`? Can they be easily removed or are there many dependencies? (I have not yet looked).

Comment (redgoat650) [reply snapshot/snapshotfs/upload.go]: you know what, I added this as the very first shot at a fix, never used it anywhere, and forgot to take it out. I can either add it as another condition for ignoring directories, or remove it.

`IgnoreFileErrors` is set in one place currently, `command_snapshot_migrate.go`.

File: snapshot/snapshotfs/upload.go (hunk 1 above)
Comment (redgoat650) [reply snapshot/snapshotfs/upload.go]: Tried that first, won't work though. Any error would be fine to be type asserted as a DirReadError.

https://play.golang.org/p/2iQD3e_Ewgd

Comment (julio-lopez) [reply snapshot/snapshotfs/upload.go]: Oh! I see. Interesting.

**Comment by redgoat650** [comment]
CI failed with `repository_connect_test.go:28: 'kopia repo create filesystem --path /tmp/kopia-repo662689381' failed with read |0: file already closed`

**Comment by redgoat650** [comment]
@julio-lopez I'm going to merge this with the base PR https://github.com/kastenhq/kopia/pull/34. If you can give the final 👍 there, I'll open the upstream PR

**Comment by julio-lopez** [comment]
> Okay I think this is ready again for review @pavannd1 @julio-lopez. Please pay special attention to the test case expected outcomes when you review, make sure you agree with the expected behavior.
> 
> I have implemented it such that the policy ignores read errors ONLY on child entries of the source. So a snapshot will still fail if the source root directory itself can't be read, but you can ignore the error if a file or a subdirectory in the snapshot source root can't be read. I did this to address some otherwise strange behavior where you would successfully snapshot (because you ignored the error), but couldn't restore that snapshot because nothing really happened during the operation.

Makes sense. This ^^^ should be included in the description when pushing upstream.
